### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/QuadraticForm/TensorProduct): base change of quadratic forms

### PR DESCRIPTION
**Pull request by eric-wieser**
---

- [x] depends on: #6306

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
I also tried to prove:
```lean
theorem _root_.BilinForm.toQuadraticForm_tmul (B₁ : BilinForm A M₁) (B₂ : BilinForm R M₂) :
    (B₁.tmul B₂).toQuadraticForm
      = B₁.toQuadraticForm.tmul B₂.toQuadraticForm := by
  letI : Invertible (2 : A) := (Invertible.map (algebraMap R A) 2).copy 2 (map_ofNat _ _).symm
  rw [QuadraticForm.tmul, tensorDistrib, BilinForm.tmul]
  dsimp
  refine (associated_rightInverse R).injective ?_
  refine BilinForm.toLin.injective ?_
  ext a₁ a₂ b₁ b₂
  dsimp [-associated_apply]
  simp [sub_smul, smul_sub, add_smul, smul_add, ←smul_mul_smul]
  set aa₁ := B₁ a₁ a₁
  set ab₁ := B₁ a₁ b₁
  set ba₁ := B₁ b₁ a₁
  set bb₁ := B₁ b₁ b₁
  set aa₂ := B₂ a₂ a₂
  set ab₂ := B₂ a₂ b₂
  set ba₂ := B₂ b₂ a₂
  set bb₂ := B₂ b₂ b₂
  sorry
```
though the algebra is too painful to get to a point where it's obvious whether it's true.

File: Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+variable (R A) in
+/-- The tensor product of two quadratic forms injects into quadratic forms on tensor products.
+
+Note this is heterobasic; the quadratic form on the left can take values in a larger ring than
+the one on the right. -/
+def tensorDistrib : QuadraticForm A M₁ ⊗[R] QuadraticForm R M₂ →ₗ[A] QuadraticForm A (M₁ ⊗[R] M₂) :=
+  letI : Invertible (2 : A) := (Invertible.map (algebraMap R A) 2).copy 2 (map_ofNat _ _).symm
+  let toQ := BilinForm.toQuadraticFormLinearMap A A (M₁ ⊗[R] M₂)
```
Comment (alexjbest) [inline Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean:42]: Would these have (marginally) cleaner definitional properties as `letI`'s instead?

File: Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+## Main definitions
+
+* `QuadraticForm.tensorDistrib (Q₁ ⊗ₜ Q₂)`: the quadratic form on `M₁ ⊗ M₂` constructed by applying
+  `Q₁` on `M₁` and `Q₂` on `M₂`. This construction is not available in characteristic two.
+
+-/
+
+universe uι uR uA uM₁ uM₂
+
+variable {ι : Type uι} {R : Type uR} {A : Type uA} {M₁ : Type uM₁} {M₂ : Type uM₂}
```
Comment (alexjbest) [inline Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean:21]:
```suggestion
universe uR uA uM₁ uM₂

variable {R : Type uR} {A : Type uA} {M₁ : Type uM₁} {M₂ : Type uM₂}
```
it doesn't seem iota is used in this file?

File: Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean:42]: These lets are indeed here as an attempt to fight timeouts

File: Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean:21]: Oops!

File: Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean:42]: Looks like `let` is 1% more heartbeats but 33% more wall time, so I'd prefer to leave this as is.

**Comment by alexjbest** [comment]
maintainer merge

File: Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean (hunk 1 above)
Comment (ocfnash) [reply Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean:42]: > These lets are indeed here as an attempt to fight timeouts

I suggest adding a comment / remark documenting that performance has impacted style here.

**Review by ocfnash — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/TensorProduct.lean:42]:
```suggestion
  -- while `letI`s would produce a better term than `let`, they would make this already-slow
  -- definition even slower.
  let toQ := BilinForm.toQuadraticFormLinearMap A A (M₁ ⊗[R] M₂)
```

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**